Fix extractSerpent2GCs.py and moltres_xs.py, and add tests

smpark7 opened the pull request:
Resolves #108 . Resolves #119 .

This PR tackles several bugs and introduces tests for the extractSerpent2GCs.py and moltres_xs.py scripts for extracting group constant XS data from Serpent 2 and/or SCALE.

1. moltres_xs.py now works with Serpent 2 data. The previous version imported the wrong number of data values for BETA_EFF and LAMBDA and triggered a mooseError when Moltres checked the length against the number of precursor groups.

2. Both extractSerpent2GCs.py and moltres_xs.py now correctly extract CHI_P data instead of CHI_T data for the prompt neutron spectra. Both GenericMoltresMaterial.C and MoltresJsonMaterial.C now use CHI_P data by default, but are also backward-compatible for old input files that still rely on CHI_T.

3. MoltresJsonMaterial.C now works in SubApps for looping precursors. In these cases, Moltres automatically prints a mooseInfo message to inform/warn users that no neutron groups were initialized in the outer loop SubApp.

4. I created tests for extractSerpent2GCs.py and moltres_xs.py using the SCALE XS data from the PinCell tutorial, and Serpent 2 XS data from a simplified molten salt fast reactor model. The test for moltres_xs.py tests the script on both SCALE and Serpent 2 data and compares the results with the gold copies. The Serpent 2 XS data gold copies from both scripts are quantitatively identical as they're parsed from the same Serpent 2 simulation.

pep8speaks commented:
Hello @smpark7! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`python/test/test_extractSerpent2GCs.py`](https://github.com/arfc/moltres/blob/64359b80645afe2a5c8a69d607500f5024e67d8c/python/test/test_extractSerpent2GCs.py):

> [Line 1:11](https://github.com/arfc/moltres/blob/64359b80645afe2a5c8a69d607500f5024e67d8c/python/test/test_extractSerpent2GCs.py#L1): [E401](https://duckduckgo.com/?q=pep8%20E401) multiple imports on one line
> [Line 7:1](https://github.com/arfc/moltres/blob/64359b80645afe2a5c8a69d607500f5024e67d8c/python/test/test_extractSerpent2GCs.py#L7): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 28:1](https://github.com/arfc/moltres/blob/64359b80645afe2a5c8a69d607500f5024e67d8c/python/test/test_extractSerpent2GCs.py#L28): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 31:1](https://github.com/arfc/moltres/blob/64359b80645afe2a5c8a69d607500f5024e67d8c/python/test/test_extractSerpent2GCs.py#L31): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 65:1](https://github.com/arfc/moltres/blob/64359b80645afe2a5c8a69d607500f5024e67d8c/python/test/test_extractSerpent2GCs.py#L65): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

* In the file [`python/test/test_moltres_xs.py`](https://github.com/arfc/moltres/blob/64359b80645afe2a5c8a69d607500f5024e67d8c/python/test/test_moltres_xs.py):

> [Line 6:11](https://github.com/arfc/moltres/blob/64359b80645afe2a5c8a69d607500f5024e67d8c/python/test/test_moltres_xs.py#L6): [E401](https://duckduckgo.com/?q=pep8%20E401) multiple imports on one line
> [Line 8:1](https://github.com/arfc/moltres/blob/64359b80645afe2a5c8a69d607500f5024e67d8c/python/test/test_moltres_xs.py#L8): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 20:35](https://github.com/arfc/moltres/blob/64359b80645afe2a5c8a69d607500f5024e67d8c/python/test/test_moltres_xs.py#L20): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 61:1](https://github.com/arfc/moltres/blob/64359b80645afe2a5c8a69d607500f5024e67d8c/python/test/test_moltres_xs.py#L61): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

smpark7 commented:
The CIVET tests are failing because of whitespaces and tabs. I'll ping everyone after I resolve them.

smpark7 commented:
Okay this is ready for review

smpark7 commented:
@robfairh With this PR, all cases will use CHIP unless a user uses old input files with old XS data. The new python scripts will no longer extract CHIT data. 

Do you think there are any cases in which CHIT is relevant/necessary?

katyhuff commented on python/extractSerpent2GCs.py:
```diff
@@ -29,7 +32,7 @@ def makePropertiesDir(
 
     # the constants moltres looks for:
-    goodStuff = ['BETA_EFF', 'Chit', 'Chid', 'lambda', 'Diffcoef', 'Kappa',
+    goodStuff = ['BETA_EFF', 'Chip', 'Chid', 'lambda', 'Diffcoef', 'Kappa',
```
this is problematic.

smpark7 replied:
In the latest changes, I've added a boolean flag such that this script extracts ChiP and ChiD by default. Otherwise, it extracts ChiT.

katyhuff commented:
> Do you think there are any cases in which CHIT is relevant/necessary?
To be clear, $\Chi_T = \Chi_D + \Chi_P$.
To replace $\Chi_T$ with $\Chi_P$ is to neglect delayed neutrons.
So, unless \Chi_D is always also included, then \Chi_P is never a sufficient replacement for \Chi_T in cases involving delayed neutrons. Since Moltres was created primarily to account for the movement of delayed neutron precursors in liquid fuelled reactors, then this impacts nearly all capability-demonstrating Moltres runs.

robfairh commented:
> The new python scripts will no longer extract CHIT data.

I think CHIT should still be extracted.

> Do you think there are any cases in which CHIT is relevant/necessary?

Yes.
For eigenvalue calculations, you have two options:
1) use the neutronics kernels and the precursor kernels. Those use CHIP and CHID.

or

2) neglect the precursors and use only the neutronics kernels. This case would use CHIT.
You get to this by combining both set of equations (diffusion and precursors) and using the equation:
![85600061-1e7b7100-b61b-11ea-8aa5-d198536edcec](https://user-images.githubusercontent.com/43791150/90701753-145aa500-e24f-11ea-9c98-3db037f8dfbb.png)
This 2nd option is what the test 'nts' does.

smpark7 commented:
@katyhuff @robfairh 
> I think the issue is the use of chi_T vs chi_P in the 'account_delayed' logic.

Yes, the main issue here is that Moltres currently uses chi_T and chi_D instead of chi_P and chi_D when ```account_delayed = true```.

> So, unless \chi_D is always also included, then \chi_P is never a sufficient replacement for \chi_T in cases involving delayed neutrons.

I agree that chi_P is not a sufficient replacement for chi_T. But what actually happens in Moltres currently when we don't provide chi_D is that Moltres assumes all delayed neutrons are born in the highest energy group while using chi_T for prompt neutrons. This approach seems like a band-aid solution for users who don't have chi_D data and it errs twice by using chi_T for prompt neutrons. I believe Moltres should use chi_P in this case while continuing to assume delayed neutrons are born in the highest energy group, to support the small subset of users who wish to use Moltres and for various reasons just happen to not have chi_D data. The alternative would be to use chi_T for both prompt and delayed neutrons when chi_D is not available. I think both approaches are less than ideal but they're the best case scenarios for users who don't have chi_D data.

> Also, this second option is helpful for large systems. If you don't require to know the precursors distribution, this approach reduces the number of DoF.

It would certainly help reduce simulation time but the results would be less accurate because it ignores precursor drift. I can see that this is a non-issue for solid fuel reactors like the HTGR so maybe we could retain extracting chi_T data as well. We may need to introduce another flag in ```CoupledFissionEigenKernel``` to tell Moltres whether to use (chi_T) or (chi_P and chi_D).

smpark7 commented:
I also want to note that the changes in this PR only affect how we extract and use Serpent 2 group constant data. Since Serpent 2 always provides chi_D data, I think it is safe to replace chi_T with chi_P. 

The group constant data that Moltres uses/requires from SCALE are ```CHI``` and ```CHID```. I'm not sure whether ```CHI``` from SCALE is equivalent to chi_T or chi_P from Serpent 2. I did not make any changes to how Moltres handles SCALE group constants.

robfairh commented:
> Yes, the main issue here is that Moltres currently uses chi_T and chi_D instead of chi_P and chi_D when `account_delayed = true`.

Actually, it does it all the time, regardless of the state of `account_delayed`.
`GenericMoltresMaterial` doesn't use the flag `account_delayed` at all.

> It would certainly help reduce simulation time but the results would be less accurate because it ignores precursor drift. I can see that this is a non-issue for solid fuel reactors like the HTGR so maybe we could retain extracting chi_T data as well. We may need to introduce another flag in `CoupledFissionEigenKernel` to tell Moltres whether to use (chi_T) or (chi_P and chi_D).

Yes,  I agree with you. That sounds like a good solution.
However, `GenericMoltresMaterial` needs to define chi = chiT.
Currently, this is the logic in `CoupledFissionKernel`:

```
Real
CoupledFissionKernel::computeQpResidual()
{
  Real r = 0;
  for (unsigned int i = 0; i < _num_groups; ++i)
    r += -_nsf[_qp][i] * computeConcentration((*_group_fluxes[i]), _qp);

  if (_account_delayed)
    r *= (1. - _beta[_qp]);

  return _test[_i][_qp] * _chi[_qp][_group] * r;
}
```

I think something like this would do what you are suggesting:
```
Real
CoupledFissionKernel::computeQpResidual()
{
  Real r = 0;
  for (unsigned int i = 0; i < _num_groups; ++i)
    r += -_nsf[_qp][i] * computeConcentration((*_group_fluxes[i]), _qp);

  if (_account_delayed)
 {
    r *= _chi_p[_qp][_group] * (1. - _beta[_qp]);
 }
 else
{
    r *= _chi_t[_qp][_group];
}
  return _test[_i][_qp] * r;
}
```
But chi_T should be added to `GenericMoltresMaterial`.

Another way to solve it would be leaving everything as it is. `CoupledFissionKernel` and use your new implementation of `GenericMoltresMaterial` but document somewhere that for using such implementation (the one that uses chiT instead of chiP and chiD) the user shouldn't have chiP among their properties.
What do you think? @smpark7

katyhuff commented on src/materials/GenericMoltresMaterial.C:
```diff
@@ -77,7 +77,21 @@ GenericMoltresMaterial::GenericMoltresMaterial(const InputParameters & parameter
+    std::string file_name = property_tables_root + "CHIP" + ".txt";
+    const std::string & file_name_ref = file_name;
+    std::ifstream myfile(file_name_ref.c_str());
+    if (myfile.good())
```
What is this?

katyhuff commented on src/materials/GenericMoltresMaterial.C:
```diff
@@ -77,7 +77,21 @@ GenericMoltresMaterial::GenericMoltresMaterial(const InputParameters & parameter
+    else
+    {
+      _file_map["CHI"] = "CHIT";
+      mooseWarning("CHI_P data missing -> using CHI_T data instead for material" + _name + ". Extract XS data using latest script for CHI_P.");
```
this is character overflow.

smpark7 replied:
I've removed the overflow in the latest changes

smpark7 commented:
From our meeting yesterday, I realized that the CHI_T/CHI_P issue warrants a more thorough discussion and my PR requires more tweaks and corrections than I had initially anticipated.

I list below all relevant scenarios that I can think of when a user runs Moltres:

1. Old examples and tests

These simulations use group constant data txt files in ```$MOLTRES/property_file_dir``` that only provide CHI (SCALE) or CHIT (Serpent 2) data. There are no CHIP or CHID txt files provided here. When we run these files, Moltres currently uses CHI or CHIT as a replacement for CHIP. In the absence of CHID data, Moltres assumes all delayed neutrons are born in the fastest neutron energy group. In this PR I had introduced code that checks whether a CHIP txt file exists to retain the option of using CHIT for backward compatibility. 

2. Their own models and input files with group constant data they generate using ```extractSerpent2GCs.py``` or ```moltres_xs.py```

The current versions of ```extractSerpent2GCs.py``` and ```moltres_xs.py``` extract CHIT and CHID data from the Serpent 2/SCALE output files that a user provides. We recognize that this is wrong because ideally, CHIT should not replace CHIP. In this PR, I initially made changes to the python scripts to extract CHIP and CHID instead, and completely ignore CHIT data. However, as @robfairh mentioned, we may see some use for CHIT, which I will describe in the next scenario.

3. Neglect precursor drift and use CHIT for large systems when users want to reduce computational load

If a user does not need to know the precursor distribution in a criticality eigensolve of a static-fuel system, the user can save computational time by setting ```account_delayed = false``` and using CHIT. 

> So, unless \chi_D is always also included, then \chi_P is never a sufficient replacement for \chi_T in cases involving delayed neutrons.

@katyhuff So with this PR, I hoped to change the behavior in scenario (2) from using CHIT and CHID to using CHIP and CHID. New users who use the python scripts to extract group constant data should always have CHID available so I think it is safe to make this change. After I make further changes in this PR for scenario (3), Moltres should use CHIT only under scenarios (1) and (3).

smpark7 commented:
Based on your feedback, I've added a flag ```use_chit``` in both ```extractSerpent2GCs.py``` and ```moltres_xs.py``` that allows users to choose between (a) extracting ChiP and ChiD (default), and (b) extracting ChiT alone. I've also added more tests to test each case separately:

1. Using ```extractSerpent2GCs.py``` to extract ChiP and ChiD
2. Using ```extractSerpent2GCs.py``` to extract ChiT
3. Using ```moltres_xs.py``` to extract ChiP and ChiD
4. Using ```moltres_xs.py``` to extract ChiT

This change should make it clear that ChiT is not a replacement for ChiP but rather, a user can choose to extract ChiT data for different use cases, one example being scenario (3) in my previous comment.

robfairh commented:
The PR is looking good!
The tests passed.
Fix the PEP8 issues and I will merge!

smpark7 commented:
@robfairh Thanks for reviewing my PR! I still have some changes to make that Katy recommended, so don't merge it until she green-lights it too.

katyhuff commented:
I'll have time to look at it tomorrow!

On Fri, Sep 18, 2020 at 9:49 AM Sun Myung Park <notifications@github.com>
wrote:

> @robfairh <https://github.com/robfairh> Thanks for reviewing my PR! I
> still have some changes to make that Katy recommended, so don't merge it
> until she green-lights it too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/arfc/moltres/pull/120#issuecomment-694913268>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADAFK3KVJGKXU3VWRC2ANLSGNXOLANCNFSM4PUF5KKQ>
> .
>


-- 
http://katyhuff.github.com

katyhuff commented on python/extractSerpent2GCs.py:
```diff
@@ -29,8 +33,13 @@ def makePropertiesDir(
-    goodStuff = ['BETA_EFF', 'Chit', 'Chid', 'lambda', 'Diffcoef', 'Kappa',
+    goodStuff = ['BETA_EFF', 'lambda', 'Diffcoef', 'Kappa',
                  'Sp0', 'Nsf', 'Invv', 'Remxs', 'Fiss', 'Nubar', 'Flx']
+    if use_chit:
```
Hi @smpark7 It seems that requiring the user to specify this could be avoided, particularly since the moltres input file itself declares whether or not delayed neutron precursors will be used. 

If chit is available, it should be recorded in goodstuff. Same with chid and chip. If any of them are not available, perhaps they could be stored as None. Then, they are all available in the input data regardless of the choice made in the moltres input file in by the moltres user at runtime.

smpark7 replied:
I've changed this so that this script extracts all three CHI data, which are always available in a Serpent 2 output file. Moltres will now decide whether to use CHIP or CHIT in ```CoupledFissionKernel``` or ```CoupledFissionEigenKernel``` based on the existing ```account_delayed``` flag.

katyhuff commented on python/extractSerpent2GCs.py:
```diff
@@ -16,6 +19,7 @@ def makePropertiesDir(
         mapFile,
         secbranchFile,
         unimapFile,
+        use_chit=False,
```
I think we could skip this. See comment below.

katyhuff commented on python/extractSerpent2GCs.py:
```diff
@@ -193,6 +202,14 @@ def makePropertiesDir(
         nargs=1,
         help='File that maps material names to serpent universe')
+    parser.add_argument(
+        '--use_chit',
```
see comment above.

katyhuff commented on python/moltres_xs.py:
```diff
@@ -162,7 +162,7 @@ class serpent_xs:
     """
 
-    def __init__(self, xs_filename):
+    def __init__(self, xs_filename, use_chit):
```
Same comment as elsewhere.

katyhuff commented on python/extractSerpent2GCs.py:
```diff
@@ -1,8 +1,11 @@
 # groups, or anything like that. Also, this could be imported into
 # other python scripts if needed, maybe for parametric studies.
+# We may phase out this script in favor of the cleaner JSON format from
+# moltres_xs.py.
```
Is there any reason to keep maintiaining extractSerpent2GCs.py at this point?

smpark7 replied:
After the fixes in this and previous PRs, ```moltres_xs.py``` and ```MoltresJsonMaterial``` can now almost replace ```extractSerpent2GCs.py``` and ```GenericMoltresMaterial```. The only difference is that ```MoltresJsonMaterial```doesn't contain the least-squares and bicubic interpolation methods for group constant data that are present in ```GenericMoltresMaterial```. However, none of the input files use these interpolation methods and I'm not sure of their merits over the more popular linear, spline, or monotone_cubic methods.

I believe we should keep ```extractSerpent2GCs.py``` until we edit the tutorial and all tests to use JSON instead.

robfairh replied:
Having extractSerpent2GCs.py gives the user an additional tool for converting the group constants from Serpent.
moltres_xs.py relies on pyne while extractSerpent2GCs.py doesn't.
The installation of pyne is not trivial for all users. Such a process shouldn't hinder the use of Moltres.
Using extractSerpent2GCs.py the user can avoid those barriers.

gridley replied:
The below is wayyyyyy more lightweight than PyNE for parsing Serpent XS, if there's any interest in that. It's insanely easy to install and could quite reasonably added as a submodule. Just making sure you guys have seen it! :smile: 

PS: Not a big fan of PyNE myself...

https://serpent-tools.readthedocs.io/en/master/

smpark7 replied:
I've personally never had much trouble installing either dependency. If I recall correctly, cyclus users had issues with PyNE because of conflicts in the dependencies. Installing PyNE using conda-forge in a clean environment is hassle-free.

Perhaps we could keep both scripts for redundancy and update extractSerpent2GCs.py to output in json format.

katyhuff reviewed (changes requested):
Minor comments, aligned with our chat on Thursday.

katyhuff reviewed (approved):
Thanks!
